xarray/core/treenode.py

```diff
@@ -778,42 +805,79 @@ def _path_to_ancestor(self, ancestor: NamedNode) -> NodePath:
         return NodePath(path_upwards)
 
 
-def zip_subtrees(*trees: AnyNamedNode) -> Iterator[tuple[AnyNamedNode, ...]]:
-    """Iterate over aligned subtrees in breadth-first order.
+class TreeIsomorphismError(ValueError):
+    """Error raised if two tree objects do not share the same node structure."""
+
+
+def group_subtrees(
+    *trees: AnyNamedNode,
+) -> Iterator[tuple[str, tuple[AnyNamedNode, ...]]]:
+    """Iterate over subtrees grouped by relative paths in breadth-first order.
+
+    `group_subtrees` allows for applying operations over all nodes of a
+    collection of DataTree objects with nodes matched by their relative paths.
+
+    Example usage::
+
+        outputs = {}
+        for path, (node_a, node_b) in group_subtrees(tree_a, tree_b):
+            outputs[path] = f(node_a, node_b)
+        tree_out = DataTree.from_dict(outputs)
 
-    Parameters:
-    -----------
+    Parameters
+    ----------
     *trees : Tree
         Trees to iterate over.
 
     Yields
     ------
-    Tuples of matching subtrees.
+    A tuple of the relative path and corresponding nodes for each subtree in the
+    inputs.
+
+    Raises
+    ------
+    TreeIsomorphismError
+        If trees are not isomorphic, i.e., they have different structures.
+
+    See also
+    --------
+    DataTree.subtree
+    DataTree.subtree_with_keys
     """
     if not trees:
         raise TypeError("must pass at least one tree object")
 
     # https://en.wikipedia.org/wiki/Breadth-first_search#Pseudocode
-    queue = collections.deque([trees])
+    queue = collections.deque([("", trees)])
```

Comment:
I guess this answers my question about why `bool(path)` would ever be `False`. But why are we doing this exactly?

Reply:
We could equivalently return `""`, `"."`, `"/"` or `./` for the top-level node. Or we could make `None` a sentinel value recognized by `from_dict`.

Given that the others are relative paths, I would lean towards the strings `""` or `"."`. I guess `"."` might make more sense since it's the canonical path returned by `NodePath`?

Reply:
> We could equivalently return "", ".", "/" or ./ for the top-level node

We shouldn't return `"/"`, because to me that would indicate an absolute, not relative path. The current code in this branch seems to handle that non-root situation just fine:

```python
In [21]: dt1 = xr.DataTree.from_dict({"root/c/a": xr.Dataset({"x": 1}), "root/c/b": xr.Dataset({"x": 2})})

In [22]: dt2 = xr.DataTree.from_dict(
    ...:     {"root/c/a": xr.Dataset({"x": 10}), "root/c/b": xr.Dataset({"x": 20})}
    ...: )

In [23]: result = {}

In [24]: for path, (node1, node2) in xr.group_subtrees(dt1['root'], dt2['root']):
    ...:     print(path)
    ...:     print(node1.dataset)
    ...:     result[path] = node1.dataset + node2.dataset
    ...: 
.
<xarray.DatasetView> Size: 0B
Dimensions:  ()
Data variables:
    *empty*
c
<xarray.DatasetView> Size: 0B
Dimensions:  ()
Data variables:
    *empty*
c/a
<xarray.DatasetView> Size: 8B
Dimensions:  ()
Data variables:
    x        int64 8B 1
c/b
<xarray.DatasetView> Size: 8B
Dimensions:  ()
Data variables:
    x        int64 8B 2
```

That looks good to me - the returned paths are clearly relative to `dt1['root']`, not the actual root (`dt`). 

> I guess "." might make more sense since it's the canonical path returned by NodePath?

I think it should be `.`, because its generally nice and intuitive to match the behaviour of `pathlib` (xref https://github.com/pydata/xarray/issues/9448):

```python
In [1]: from pathlib import PurePath

In [2]: PurePath('a/b/').relative_to('a/b/')
Out[2]: PurePosixPath('.')

In [3]: PurePath('a/b/').relative_to('a/')
Out[3]: PurePosixPath('b')

In [4]: PurePath('a/').relative_to('a/b', walk_up=True)  # requires python 3.12
Out[4]: PurePosixPath('..')

In [5]: PurePath('a/b/').relative_to('.')
Out[5]: PurePosixPath('a/b')
```
`NodePath` just inherits that behaviour.

Reply:
I can also see an argument for `'./'`, because then direct string concatenation with `a/b` gives a valid relative path.